### PR TITLE
fix: handle assignments to dereferenced pointer values

### PR DESCRIPTION
**Pull request by deelawn**
Addresses #1326 

This is still a WIP and I need to do a lot more testing with different cases and tweak it accordingly.  Before I spend the time to do that, I'm requesting feedback as to whether or not this is the correct approach to fixing this issue.  I tried fixing it in a few different places prior to this PR; this was the only place I was successful in doing so.  Perhaps the reference value should be resolved and the base value set earlier; I'm not sure.

cc @jaekwon 

<details><summary>Contributors' checklist...</summary>

- [ ] Added new tests, or not needed, or not feasible
- [ ] Provided an example (e.g. screenshot) to aid review or the PR is self-explanatory
- [ ] Updated the official documentation or not needed
- [ ] No breaking changes were made, or a `BREAKING CHANGE: xxx` message was included in the description
- [ ] Added references to related issues and PRs
- [ ] Provided any useful hints for running manual tests
- [ ] Added new benchmarks to [generated graphs](https://gnoland.github.io/benchmarks), if any. More info [here](https://github.com/gnolang/gno/blob/master/.benchmarks/README.md).
</details>

### CODE REVIEW

**Comment by jaekwon** [comment]
taking a look...

File: gnovm/pkg/gnolang/values.go [hunk 1]
```diff
@@ -273,7 +273,35 @@ func (pv PointerValue) Assign2(alloc *Allocator, store Store, rlm *Realm, tv2 Ty
 		// Special case of DataByte into (base=*SliceValue).Data.
 		pv.TV.SetDataByte(tv2.GetUint8())
 		return
+	} else if rlm != nil && pv.Base == nil && pv.Index == 0 {
+		// An index of zero indicates this is an assignment to a realm
```
Comment (jaekwon) [inline gnovm/pkg/gnolang/values.go:277]: An index of zero could mean it's the zero'th element of an array, or the zero'th element of a block, 
or whatever the .Base type is. The Index doesn't say much but the .Base type says more but depends on what you mean by a realm object. Like a PackageValue?

**Comment by jaekwon** [comment]
Sidenote: This depends on https://github.com/gnolang/gno/pull/1305/files DeepCopy but DeepCopy is a misnomer.

**Comment by jaekwon** [comment]
I think this is the approach we should take: https://github.com/gnolang/gno/pull/1501/files generally speaking.

File: gnovm/pkg/gnolang/values.go (hunk 1 above)
Comment (deelawn) [reply gnovm/pkg/gnolang/values.go:278]: You're right that an index of zero can mean more than one thing depending on the context, but in this case the realm is not nil and the base is nil, so I think this means that a value is being assigned to a realm object. I'm not sure what the difference is between realm object and PackageValue; they might be synonymous -- the values that are persisted from one block to the next; variables that are declared in a realm and are not local to any function or method.

Comment (jaekwon) [reply gnovm/pkg/gnolang/values.go:278]: ok at least the comment needs fixing, "A base of nil indicates"? 
A base of nil indicates an assignment to an escaped object.
An escaped object has no owner.
All objects are realm objects.
Some are substructures (base == parent, ref==1), others are escaped.
Unless I'm confused haha. 

All realms have an associated package value but not all packages are realms. Depends on the PkgPath name.

**Comment by deelawn** [comment]
Closing in favor of #1501

**Pull request closed without merging**